### ISSUE TITLE
Instance search fails too eagerly when generalization will introduce instance

### ISSUE THREAD

**Issue by UlfNorell** [agda/agda#3358]
```agda
open import Agda.Builtin.Equality

record IsGroup (G : Set) : Set where
  field _∙_ : G → G → G

open IsGroup ⦃ ... ⦄

record Group : Set₁ where
  field G       : Set
        ⦃ IsG ⦄ : IsGroup G

open Group using () renaming (G to [_])

variable
  G : Group

postulate
  works : ∀ {G} → (x : [ G ]) → (x ∙ x) ≡ x
  fails : (x : [ G ]) → (x ∙ x) ≡ x
  -- No instance of type IsGroup [ G ]
```

**Comment by jespercockx** [comment]
I'll take a look at this.

**Comment by jespercockx** [comment]
I'm trying to fix the problem by postponing instance search in case we can't yet determine the type of some of the context variables (which may become a record type with an instance field). However, I run into two problems:

1. It breaks some test cases (such as #1952) where one of the argument types is a metavariable which is only solved when we solve the instance meta. With my change, it wouldn't be solved because the meta **could** be a record type which contains another instance, so we cannot 100% rule out the possibility of overlapping instances.

2. This doesn't actually fix the problem because generalization first creates a 'dummy' record with no fields and only updates this definition later, but by then it's too late. This seems like a quite horrible hack, I don't want to take into account the fact that there may be new fields added to a record type later. Can one of the 'generalize' specialists @UlfNorell or @divipp tell me what I should do in this situation?

**Comment by jespercockx** [comment]
I pushed my work so far to here https://github.com/jespercockx/agda/commit/f8078614b53f4cd9c3c9895903625b7489cef898

Currently it breaks the test cases `Succeed/Issue1952.agda` and `test/LibSucceed/InstanceArguments/11-monads.agda`.

**Comment by UlfNorell** [comment]
> This doesn't actually fix the problem because generalization first creates a 'dummy' record with no fields and only updates this definition later, but by then it's too late. This seems like a quite horrible hack, I don't want to take into account the fact that there may be new fields added to a record type later.

I suppose we could create an extra metavariable for the type and not create the record type until we know the fields.

**Comment by jespercockx** [comment]
I tried to merge the two functions `createGenRecordType` and `fillInGenRecordDetails` but I messed up somehow because I got strange errors from the test suite. Maybe someone else could attempt this who understands the code better?

Instead, I decided to just add a `postponeInstanceConstraints` around the whole `computeGeneralization` function, so instance search is never tried while there's a bogus record in the signature, which seems to fix the problem. Patch incoming once the tests are finished.

**Comment by jespercockx** [comment]
Here's a small puzzle:
```agda
it : {X : Set} {{_ : X}} → X
it {{x}} = x

postulate
  A : Set
  P : Set → Set

record R : Set where
  field {{fld}} : P A

postulate instance inst : P R

mutual
  ?X = _

  test : ?X → P ?X
  test = it
```
Should instance search here pick the instance `inst`? At first it seems to be the only instance in scope. However, when we pick it then `?X` is instantiated to `R`, which suddenly brings another instance `fld x` in scope! So the solution wasn't unique after all. So it seems better to postpone instance search until all context metas are solved.

**Comment by jespercockx** [comment]
I pushed my fix, with the necessary changes to the two test cases I mentioned. I looked a long time at the code for `11-monads.agda` and I'm genuinely unsure whether the constraints you get after my change ought to be solvable or not. Here is a fully explicit example that no longer works now:
```agda
{-# OPTIONS -v tc.term.args:30 -v tc.term.expr.top:15 -v tc.check.app:69 -v tc.term.def:69 -v tc.meta.blocked:69 -v tc.meta.assign:50 -v tc.conv.term:10 -v tc.constr.solve:50 #-}

it : {X : Set} {{_ : X}} → X
it {{x}} = x

postulate
  Monad : (Set → Set) → Set
  bind : (M : Set → Set) → Monad M → (A B : Set) → M A → (A → M B) → M B

  X Y : Set
  M₁ M₂ : Set → Set
  instance
    Monad-M₁ : Monad M₁
    Monad-M₂ : Monad M₂

  f : M₁ (M₂ X)
  g : M₂ X → M₁ (M₂ X)
  h : X → M₂ X

mutual
  ?M₂ : Set → Set
  ?M₂ = _

  test : M₁ (M₂ X)
  test = bind M₁ Monad-M₁ (?M₂ X) (M₂ X) f (λ x → g (bind ?M₂ it X X x h))
```
The error message is:
```
Failed to solve the following constraints:
  Resolve instance argument _19 : (x₁ : ?M₂ X) → Monad ?M₂
  No candidates yet
  _22 := λ x → bind ?M₂ (_19 (x = x)) X X x (_21 (x = x))
    [blocked on problem 49]
  [49] _15 X =< M₂ X : Set
  _20 := λ x → h [blocked on problem 47]
  [47] M₂ X =< _15 X : Set
  _16 := f [blocked on problem 38]
  [38, 39] M₂ X = _15 X : Set
```
The crucial constraint here is `M₂ X = _15 X : Set` (where `X` is a postulate). It looks like it *should* be solvable, but I'm honestly not sure.

**Comment by nad** [comment]
@jespercockx, ca4fceabfc0d70a257c94cdfc2a62222bd4ea0a6 and/or cd6cee087e889bf22d7495bd5deb77711c237b4f introduce quite a bit of breakage in my code. I've constructed a small example:
```agda
record R {A : Set} (P : A → Set) : Set where
  field
    r : ∀ x → P x → P x

open R ⦃ … ⦄

data D : Set where
  c : D → D

data E : Set where
  c : E  -- Note that c is overloaded.

postulate
  P : D → Set
  p : ∀ {x} → P (c x)

  instance
    r-P : R P

no-type-signature-needed = λ {x} →
  r (c x) p
```
The code above is no longer accepted. Is this intended and necessary?

**Comment by jespercockx** [comment]
Well, the type of `x` in the definition of `no-signature-needed` is only determined to be `D` once the instance is filled in. However it could alternatively be some record type which contains another instance of `R ...` as an instance field. Since we cannot determine for sure that the solution will be unique, the semantics of instance arguments dictate that we shouldn't solve it.

**Comment by nad** [comment]
> However, when we pick it then `?X` is instantiated to `R`, which suddenly brings another instance `fld x` in scope!

I don't understand what you mean: the instance `R.fld` is already in scope, no matter what `?X` is. However, I suppose that only makes the problem worse.

I think this change to the semantics should be documented in the changelog and the reference manual.

**Comment by jespercockx** [comment]
This is just a bugfix: the old implementation wasn't following the semantics of instance arguments, and the new one does. So I'm not sure what to document.

**Comment by nad** [comment]
OK.
